### PR TITLE
Generate uber jar for native image source jar if the uberJar option of quarkus-maven-plugin is enabled

### LINKED ISSUE #8850: Support uber jar for native image source jar
**Description**
Currently, quarkus-maven-plugin ignores the `uberJar` option for native image builds. However, using an uber jar as native image source jar can be used as a workaround for a current GraalVM [limitation ](https://github.com/oracle/graal/issues/2387) on Windows.

**Implementation ideas**
Reuse the existing uber jar creation logic in the quarkus-maven-plugin for native image source jar creation if the `uberJar` option is enabled.

### PR DESCRIPTION
**Pull request by Mobe91**
…option of quarkus-maven-plugin is enabled

fixes #8850

### CODE REVIEW

**Comment by stuartwdouglas** [comment]
I don't really want to support this if we can avoid it, it would be a lot better if this was fixed in Graal as it is simpler to only maintain a single native image format.

**Comment by Mobe91** [comment]
Understandable but this implies that Windows will remain unusable for "real" applications for several more months. And even when GraalVM fixes this eventually, it will still take some more time for Quarkus to add support for new GraalVM version. So we are talking about a 6 months+ I would say.

This is a very simple change that reuses the existing uberJar code and just makes it available for native image source building, too. So, a relatively small support overhead for instant support of "real" applications on Windows. For me, this seems like a good tradeoff.

**Comment by stuartwdouglas** [comment]
I guess I am ok with it for now as a trade off, but it should be documented that we will remove it once this particular bug is fixed and we have upgraded to the fixed Graal version.

**Comment by Mobe91** [comment]
Great 👍 
Would you like me to add this as in-code documentation?

**Comment by stuartwdouglas** [comment]
Maybe we should automatically do this if and only if we detect that we are doing a native build on windows, what do you think @gsmet ?

**Comment by gsmet** [comment]
+1 to do that automatically only on Windows for the time being.

The only thing is that this might introduce more Windows-specific issues.

**Comment by geoand** [comment]
Also +1 for doing the switch for Windows, but there should be a good warning message that this was done.

**Comment by Mobe91** [comment]
@stuartwdouglas Anything else missing for this to be merged?

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java [hunk 1]
```diff
@@ -593,23 +616,54 @@ public NativeImageSourceJarBuildItem buildNativeImageJar(CurateOutcomeBuildItem
...
-        Files.createDirectories(libDir);
+        IoUtils.recursiveDelete(targetDirectory);
+        Files.createDirectories(targetDirectory);
 
         List<GeneratedClassBuildItem> allClasses = new ArrayList<>(generatedClasses);
         allClasses.addAll(nativeImageResources.stream()
                 .map((s) -> new GeneratedClassBuildItem(true, s.getName(), s.getClassData()))
                 .collect(Collectors.toList()));
 
+        boolean uberJarRequested = !uberJarRequired.isEmpty() || packageConfig.uberJar;
```
Comment (stuartwdouglas) [inline core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java:631]: I think this should only be for windows. We just want one format to build the native image from in general, this is an intermediate state so it should not matter to the user, but having two different formats means more possible things to debug.

**Review by stuartwdouglas — CHANGES_REQUESTED** [review]
I think this should only be for the windows workaround

File: core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java (hunk 1 above)
Comment (Mobe91) [reply core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/JarResultBuildStep.java]: @stuartwdouglas Can you check if my last change is what you had in mind?

**Review by stuartwdouglas — APPROVED** [review]

**Comment by gsmet** [comment]
I won't backport that one as `JarResultBuildStep` got basically rewritten with the reaugment stuff so it's not a simple backport.

This will have to wait for 1.6.

**Comment by zakkak** [comment]
https://github.com/oracle/graal/issues/2387 has now been fixed and the fix is present in Graal VM CE 21.0.0, 20.3.1, 20.3.0.
I think it is safe to revert this, but I can't test it.

**Comment by shumonsharif** [comment]
Are there any plans to revert these changes, given the issue has been fixed in GraalVM?  Kindly refer to https://github.com/quarkiverse/quarkus-cxf/issues/238 - it appears as though the native image source jar causes issues for our extension on Windows only.

**Comment by Sanne** [comment]
> [oracle/graal#2387](https://github.com/oracle/graal/issues/2387) has now been fixed and the fix is present in Graal VM CE 21.0.0, 20.3.1, and 20.3.0. I think it is safe to revert this, but I can't test it.

@zakkak @shumonsharif This seems to have fallen through the cracks, following up with #26289

**Pull request merged**